deeppavlov/pipeline_manager/pipeline_manager.py

```diff
@@ -0,0 +1,424 @@
+        self.save_best = self.exp_config['pipeline_search'].get('save_best', False)
+        self.search_type = self.exp_config['pipeline_search'].get('search_type', 'random')
+        self.sample_num = self.exp_config['pipeline_search'].get('sample_num', 10)
+        self.target_metric = self.exp_config['pipeline_search'].get('target_metric')
+        self.multiprocessing = self.exp_config['pipeline_search'].get('multiprocessing', False)
+        self.num_workers = self.exp_config['pipeline_search'].get('num_workers')
+        self.use_gpu = self.exp_config['pipeline_search'].get('use_gpu', False)
+        self.memory_fraction = self.exp_config['pipeline_search'].get('gpu_memory_fraction', 0.99)
+        self.available_gpu = None
+
+        # create the observer
+        launch_name = self.exp_config['pipeline_search'].get('launch_name', 'exp')
+        self.observer = ExperimentObserver(self.exp_name, launch_name, self.root_path, self.info, self.date)
+        # create the pipeline generator
+        self.pipeline_generator = PipeGen(self.exp_config, self.observer.save_path, self.search_type, self.sample_num)
+        self.generated_configs = [conf for conf in self.pipeline_generator()]
+        self.gen_len = len(self.generated_configs)
+
+        # write train data in observer
+        self.observer.exp_info['number_of_pipes'] = self.gen_len
+        self.observer.exp_info['experiment_config'] = str(config_path)
+
+        self.add_metric_info()
+        if self.use_gpu is not False:
+            self.gpu_preparation()
+
+        # write time of experiment start
+        self.start_exp = time.time()
+        # start test
+        if self.exp_config['pipeline_search'].get('do_test', False):
+            self.test()
+
+    def add_metric_info(self) -> None:
+        """ Adding metrics information in observer. """
+        self.observer.exp_info['metrics'] = []
+        for met in self.exp_config['train']['metrics']:
+            if isinstance(met, dict):
+                self.observer.exp_info['metrics'].append(met['name'])
+            else:
+                self.observer.exp_info['metrics'].append(met)
+
+        if self.target_metric:
+            self.observer.exp_info['target_metric'] = self.target_metric
+        else:
+            self.observer.exp_info['target_metric'] = self.observer.exp_info['metrics'][0]
+
+    def gpu_preparation(self) -> None:
+        """
+        Calculates the number of workers and the set of available video cards, if gpu is used, based on init attributes.
+        """
+        try:
+            cuda_vis_dev = os.environ['CUDA_VISIBLE_DEVICES']
+            if cuda_vis_dev != '':
+                visible_gpu = [int(q) for q in cuda_vis_dev.split(',')]
+                os.environ['CUDA_VISIBLE_DEVICES'] = ""
+            else:
+                visible_gpu = None
+        except KeyError:
+            visible_gpu = None
+
+        if isinstance(self.use_gpu, (list, int)):
```

Comment:
yoptar, you said that "This is not how CUDA_VISIBLE_DEVICES should work". I don't understand what you mean.